packages/components/src/navigation/children-utils.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * External dependencies
+ */
+import { keyBy, omit } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { ROOT_MENU } from './constants';
+import NavigationItem from './item';
+import NavigationMenu from './menu';
+
+/**
+ *
+ * @typedef {Object} ChildWithClosestParent
+ * @property {WPElement} child element that matched the predicate
+ * @property {WPElement?} parent closest matching parent that matched the parentPredicate
+ */
+
+/**
+ *
+ * @param {WPElement} initialChildren children to iterate through recursively
+ * @param {Function} predicate children matching this function predicate will be returned
+ * @param {Function} parentPredicate closest parent will be determined by this function predicate
+ *
+ * @return {ChildWithClosestParent[]} children with their closest parent
+ */
+export const findChildrenWithClosestParent = (
+	initialChildren,
+	predicate,
+	parentPredicate
+) => {
+	const hasPredicate = typeof predicate === 'function';
+	const hasParentPredicate = typeof parentPredicate === 'function';
+
+	const innerRecursion = ( children, lastParent ) => {
+		let items = [];
+
+		if ( ! Array.isArray( children ) ) {
+			children = [ children ];
+		}
```

Comment:
I've just realized we _do_ expose [`React.Children`](https://reactjs.org/docs/react-api.html#reactchildren) 🤦 :
`import { Children } from '@wordpress/element';`

By using that we can avoid doing this conversion, and we can just loop with `Children.forEach()` (or `Children.toArray()` if you'd rather keep the `for`)